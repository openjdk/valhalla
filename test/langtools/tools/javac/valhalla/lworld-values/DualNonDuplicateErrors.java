/*
 * @test /nodynamiccopyright/
 * @bug 8267843
 * @summary Check that javac diagnoses `this` being passed around and instance method being invoked before primitive class instance is fully initialized.
 * @compile/fail/ref=DualNonDuplicateErrors.out -XDrawDiagnostics DualNonDuplicateErrors.java
 */

public primitive class DualNonDuplicateErrors {

    int x;

    DualNonDuplicateErrors() {
        // The call below should trigger two errors - they are not duplicates really.
        // First one is for `this` being passed around ("exposed")
        // Second is for instance method being invoked thereby allowing that method to
        // observe the primitive class instance in a partially initialized state.
        foo(this);
        x = 10;
        foo(this); // No error here.
    }

    void foo(DualNonDuplicateErrors x) {
    }
}
