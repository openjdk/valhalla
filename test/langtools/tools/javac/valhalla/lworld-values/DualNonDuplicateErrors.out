DualNonDuplicateErrors.java:17:13: compiler.err.this.exposed.prematurely
DualNonDuplicateErrors.java:17:12: compiler.err.this.exposed.prematurely
2 errors
